text/0626-invoke-helper.md

```diff
@@ -0,0 +1,475 @@
+- Start Date: 2020-04-30
+- Relevant Team(s): Ember.js
+- RFC PR: https://github.com/emberjs/rfcs/pull/626
+- Tracking: (leave this empty)
+
+# JavaScript Helper Invocation API
+
+## Summary
+
+This RFC proposes a new API, `invokeHelper`, which can be used to invoke a
+helper definition, creating an instance of the helper in JavaScript.
+
+```js
+// app/components/data-loader.js
+import Component from '@glimmer/component';
+import Helper from '@ember/component/helper';
+import { invokeHelper } from '@ember/helper';
+
+class FetchTask {
+  @tracked isLoading = true;
+  @tracked isError = false;
+  @tracked result = null;
+
+  constructor(url) {
+    this.run(url);
+  }
+
+  async run(url) {
+    try {
+      let response = await fetch(url);
+      this.result = await response.json();
+    } catch {
+      this.isError = true;
+    } finally {
+      this.isLoading = false;
+    }
+  }
+}
+
+class RemoteData extends Helper {
+  compute([url]) {
+    return new FetchTask(url);
+  }
+}
+
+export default class DataLoader extends Component {
+  data = invokeHelper(this, RemoteData, () => {
+    positional: [this.args.url]
```

Comment:
This is awesome, I've needed this many times. When I first looked at this, I expected it to be a little closer to the way the helper looks in a template (`{{remote-data 'https://google.com'}}`). I think I understand that that's not entirely possible given the nature of tracked properties (and also passing these params at instantiation).

I wonder if something like this would work?

```javascript
get data() {
  return invokeHelper(this, RemoteData, [this.args.url], /* { } */);
}
```

That might get around the tracked properties issue, but then the helper would be invoked each time.

Is "invoke" the right word for this? "bindHelper"?

Reply:
> That might get around the tracked properties issue, but then the helper would be invoked each time.

Right, exactly. You would need to do something like:

```js
get data() {
  return invokeHelper(this, 'data', RemoteData, [this.args.url] /*, { } */);
}
```

Where the second argument is a unique key specifically for this instance. Overall this API would be much less flexible, which is why I proposed the one in this RFC. We can expose higher level APIs built on top of the low level one that work more like the way you described.

> Is "invoke" the right word for this? "bindHelper"?

Binding typically doesn't create an instance of something, it curries it and returns a new definition. This actually is creating an instance, so invoking is more correct I believe.